src/client/auth/scram.rs

```diff
@@ -304,18 +363,22 @@ fn parse_kvp(str: &str, expected_key: char) -> Result<String> {
 }
 
 /// Model of the first message sent by the client.
-struct ClientFirst {
+#[derive(Debug)]
+pub(crate) struct ClientFirst {
     source: String,
 
     message: String,
 
     gs2_header: Range<usize>,
 
     bare: Range<usize>,
+
+    nonce: String,
```

Comment:
One quirk in adapting the existing handshake algorithm to support speculative auth is that the nonce is generated before the `authenticate_stream` function is called. Since the `ClientFirst` message already needs to be passed in, the simplest way to pass in the nonce is just to piggyback it on the ClientFirst. In terms of space efficiency, this is identical to passing it in as a separate argument, and since the only ways ClientFirst is used is to pass into this module, it's not present in any cases where it's not being used.